### PR TITLE
Only dispatch livelink clicks to views with a router

### LINKED ISSUE #404: Combining views in layout with routed liveviews can not use live_link
### Environment

* Elixir version (elixir -v): Elixir 1.9.1 
* Phoenix version (mix deps): 1.4.10
* NodeJS version (node -v): v8.11.1
* NPM version (npm -v):5.6.0
* Operating system: MacOS Mojave

### Actual behavior

We include a small widget made with LiveView within a layout so we can use it on all pages.
However, if we then use a full LiveView coming from an route we can not use live_link inside that liveview (which basically for us means there is no benefit for liveview anymore). 

If we would click a link in the main view the other live view crashes with a message saying: 
`** (ArgumentError) cannot invoke handle_params/3 on MultiLiveviewBugWeb.ClockLive because it is not mounted nor accessed through the router live/3 macro`

This is not expected since there is no parent child relationship between these views, they should be able to operate on their own.

I have created a small demo app that demonstrates this bug:
https://github.com/Wijnand/multi_liveview_bug
After starting you only need to visit http://localhost:4000/data and click the "filter" links.
The timer stops refreshing and the error is in the console.

### Expected behavior
I expect the live_link to only touch the liveview it is embedded in (potentially trickling down to its children). Liveviews that are standalone should not be affected.

### PR DESCRIPTION
**Pull request by Wijnand**
Since there basically can only be one main view with a router attached
it does not make sense to call handle_params on other non router views.

With this commit we mark this special view with a data attribute called
phxMain and we only push internal links
to this view instead of the root view

Fixes #404

### CODE REVIEW

**Comment by Wijnand** [comment]
I will fix the test on Monday, i want to take a somewhat different approach anyway.

**Comment by Wijnand** [comment]
@josevalim This MR fixes the issues I have with multiple LiveViews on the same page. Is this the direction you intended ?

File: lib/phoenix_live_view/view.ex [hunk 1]
```diff
@@ -347,10 +347,16 @@ defmodule Phoenix.LiveView.View do
 
     case call_mount_and_handle_params!(socket, router, view, session, conn.params, request_url) do
       {:ok, socket} ->
+        data_attrs = [
+          phx_view: config.name,
+          phx_session: sign_root_session(socket, router, view, session)
+        ]
+
+        data_attrs = if router, do: Keyword.put(data_attrs, :phx_main, true), else: data_attrs
```
Comment (josevalim) [inline lib/phoenix_live_view/view.ex]:
```suggestion
        data_attrs = (if router, do: [phx_main: true], else: []) ++ data_attrs
```

**Comment by josevalim** [comment]
Awesome work @Wijnand! :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

I had just a minor nitpick on the Elixir code. However, we will need to wait until someone more capable can review the JS bits.

**Comment by Wijnand** [comment]
@josevalim I noticed that this PR broke with recent changes in master and I updated the implementation to cope with it.

What can we do to move this forward? We only need someone to take the time for reviewing the JS bits ?

**Comment by chrismccord** [comment]
❤️❤️❤️🐥🔥

**Pull request merged**